auth/impersonate/impersonate.go

```diff
@@ -0,0 +1,208 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package impersonate
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	"cloud.google.com/go/auth"
+	"cloud.google.com/go/auth/detect"
+	"cloud.google.com/go/auth/httptransport"
+	"cloud.google.com/go/auth/internal"
+)
+
+var (
+	iamCredentialsEndpoint = "https://iamcredentials.googleapis.com"
+	oauth2Endpoint         = "https://oauth2.googleapis.com"
+)
+
+// NewCredentialTokenProvider returns an impersonated
+// [cloud.google.com/go/auth/TokenProvider] configured with the provided options
+// and using credentials loaded from Application Default Credentials as the base
+// credentials if not provided with the opts.
+func NewCredentialTokenProvider(opts *CredentialOptions) (auth.TokenProvider, error) {
+	if opts == nil {
+		return nil, fmt.Errorf("impersonate: opts must be provided")
+	}
+	if opts.TargetPrincipal == "" {
+		return nil, fmt.Errorf("impersonate: a target service account must be provided")
+	}
+	if len(opts.Scopes) == 0 {
+		return nil, fmt.Errorf("impersonate: scopes must be provided")
+	}
+	if opts.Lifetime.Hours() > 12 {
+		return nil, fmt.Errorf("impersonate: max lifetime is 12 hours")
+	}
+
+	var isStaticToken bool
+	// Default to the longest acceptable value of one hour as the token will
+	// be refreshed automatically if not set.
+	lifetime := 3600 * time.Second
```

Comment:
Why not just `1 * time.Hour`?

Reply:
I think I had done this because the resulting api request is in seconds, but you are right. 1 hour is more readable.